pkg/cli: allow tsdump to also dump the labeled child metrics

jasonlmfong opened the pull request:
This adds the ability to dump the child metrics which was introduced in https://github.com/cockroachdb/cockroach/pull/158148.
These child metrics are discovered in the tsdump through their unique resolution
and a limit list of qualified metric names.

Epic: CRDB-55079
Release: None

aa-joshi reviewed (commented):
Lets make sure that integrity of tsdump upload is not getting here: https://github.com/cockroachdb/cockroach/blob/1f9c4adf128707141a9012760c101645f405d3e0/pkg/cli/tsdump_upload.go#L709

> @aa-joshi reviewed 4 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-:-OfyLM8R-daUWdkrjUKM:b38lppp)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Abhinav1299, @angles-n-daemons, @arjunmahishi, and @jasonlmfong)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyJJfF79tlAXwQfs0c:-OfyJJfF79tlAXwQfs0d:b-l1awcm):*
Let's add reference to PR where we have updated child metrics persistence and details around how are querying metrics.
___
*[`pkg/ts/server.go` line 587 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyKcRyAblFqat9cL1-:-OfyKcRyAblFqat9cL10:b-e5zff7) ([raw file](https://github.com/cockroachdb/cockroach/blob/676c125a74776fad49a194e3c3a63cd3f34465d4/pkg/ts/server.go#L587)):*
> ```Go
> 			if includeChildMetrics {
> 				// Dump everything for this metric prefix
> 				if err := dump(&resp.Rows[i]); err != nil {
> ```

Do we need to dump parent and child metrics in includeChildMetrics? If we know that sum of all child metrics represents the parent metrics then I believe we can omit dumping parent metric to avoid doubling the value on aggregation. PR reference around similar issue: https://github.com/cockroachdb/cockroach/pull/141601
___
*[`pkg/ts/server.go` line 596 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyKEyM5Fsvs01_tlAa:-OfyKEyM5Fsvs01_tlAb:b-rijptl) ([raw file](https://github.com/cockroachdb/cockroach/blob/676c125a74776fad49a194e3c3a63cd3f34465d4/pkg/ts/server.go#L596)):*
> ```Go
> 					return err
> 				}
> 				if !strings.Contains(name, "{") {
> ```

Is there any better way to differentiate between parent and child metric?
___
*[`pkg/cli/debug.go` line 1627 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyJjfq4OZzZYohn60g:-OfyJjfq4OZzZYohn60h:b-e6v96h) ([raw file](https://github.com/cockroachdb/cockroach/blob/676c125a74776fad49a194e3c3a63cd3f34465d4/pkg/cli/debug.go#L1627)):*
> ```Go
> 	f.Int64Var(&debugTimeSeriesDumpOpts.ddMetricInterval, "dd-metric-interval", debugTimeSeriesDumpOpts.ddMetricInterval, "interval in seconds for datadoginit format only (default 10). Regular datadog format uses actual intervals from tsdump.")
> 	f.Lookup("dd-metric-interval").Hidden = true // this is for internal use only
> 	f.BoolVar(&debugTimeSeriesDumpOpts.includeChildMetrics, "include-child-metrics", false, "include changefeed child metrics with labels by using prefix pattern scanning")
> ```

Would it be possible to remove word `changefeed` from flag description?


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158716)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @angles-n-daemons, and @arjunmahishi)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyJJfF79tlAXwQfs0c:-OfyOZ_f5PZS1Wx6bkji:b-80tlqk):*
<details><summary><i>Previously, aa-joshi (Akshay Joshi) wrote…</i></summary><blockquote>

Let's add reference to PR where we have updated child metrics persistence and details around how are querying metrics.
</blockquote></details>

Sure.
___
*[`pkg/cli/debug.go` line 1627 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyJjfq4OZzZYohn60g:-OfyOe7t1MwJppe77Q5D:bxxx4gf) ([raw file](https://github.com/cockroachdb/cockroach/blob/676c125a74776fad49a194e3c3a63cd3f34465d4/pkg/cli/debug.go#L1627)):*
<details><summary><i>Previously, aa-joshi (Akshay Joshi) wrote…</i></summary><blockquote>

Would it be possible to remove word `changefeed` from flag description?
</blockquote></details>

That makes sense.
___
*[`pkg/ts/server.go` line 587 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyKcRyAblFqat9cL1-:-OfyOkf9BkN1A5VMss1R:bd884e8) ([raw file](https://github.com/cockroachdb/cockroach/blob/676c125a74776fad49a194e3c3a63cd3f34465d4/pkg/ts/server.go#L587)):*
<details><summary><i>Previously, aa-joshi (Akshay Joshi) wrote…</i></summary><blockquote>

Do we need to dump parent and child metrics in includeChildMetrics? If we know that sum of all child metrics represents the parent metrics then I believe we can omit dumping parent metric to avoid doubling the value on aggregation. PR reference around similar issue: https://github.com/cockroachdb/cockroach/pull/141601
</blockquote></details>

I think that does make sense, but currently they are dumped in separate iterations of `dumpTimeseriesAllSources()`
Let me consider how this can work
___
*[`pkg/ts/server.go` line 596 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyKEyM5Fsvs01_tlAa:-OfyPVCuAONpnA1HcUTQ:b-hbi7t3) ([raw file](https://github.com/cockroachdb/cockroach/blob/676c125a74776fad49a194e3c3a63cd3f34465d4/pkg/ts/server.go#L596)):*
<details><summary><i>Previously, aa-joshi (Akshay Joshi) wrote…</i></summary><blockquote>

Is there any better way to differentiate between parent and child metric?
</blockquote></details>

I'm not sure here, since these are just names in the KV layer.


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
I don't really understand what you mean here, can you elaborate?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-:-OfzOSkuA8Ly8XrkiIU5:b1dolt0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @angles-n-daemons, and @arjunmahishi)

___
*[`pkg/ts/server.go` line 587 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyKcRyAblFqat9cL1-:-OfyknBoDsbcOnj7oHXo:bnjum4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/676c125a74776fad49a194e3c3a63cd3f34465d4/pkg/ts/server.go#L587)):*
<details><summary><i>Previously, jasonlmfong (Jason Fong) wrote…</i></summary><blockquote>

I think that does make sense, but currently they are dumped in separate iterations of `dumpTimeseriesAllSources()`
Let me consider how this can work
</blockquote></details>

I gave it some more thought, it can still be beneficial to include the parent metrics, especially since the child metrics are stored at a lower resolution, what do you think?
___
*[`pkg/ts/server.go` line 596 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyKEyM5Fsvs01_tlAa:-OfzOJaY47-bCkC1cZC7:bg5e01k) ([raw file](https://github.com/cockroachdb/cockroach/blob/676c125a74776fad49a194e3c3a63cd3f34465d4/pkg/ts/server.go#L596)):*
<details><summary><i>Previously, jasonlmfong (Jason Fong) wrote…</i></summary><blockquote>

I'm not sure here, since these are just names in the KV layer.
</blockquote></details>

I gave it more thought and refactored this code, the code now splits on dumpImpl and we no longer need this check


<!-- Sent from Reviewable.io -->

aa-joshi reviewed (commented):
> @aa-joshi reviewed 4 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158716)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Abhinav1299, @angles-n-daemons, @arjunmahishi, and @jasonlmfong)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyJJfF79tlAXwQfs0c:-Og2Pa5I3ijTW1zn6q8E:b8k6fos):*
<details><summary><i>Previously, jasonlmfong (Jason Fong) wrote…</i></summary><blockquote>

Sure.
</blockquote></details>

Let's format the git message and I believe the details around querying metrics is still missing.
___
*[`pkg/ts/server.go` line 587 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyKcRyAblFqat9cL1-:-Og2T7BnCwpwXNsfJZt8:bwjiqt1) ([raw file](https://github.com/cockroachdb/cockroach/blob/676c125a74776fad49a194e3c3a63cd3f34465d4/pkg/ts/server.go#L587)):*
<details><summary><i>Previously, jasonlmfong (Jason Fong) wrote…</i></summary><blockquote>

I gave it some more thought, it can still be beneficial to include the parent metrics, especially since the child metrics are stored at a lower resolution, what do you think?
</blockquote></details>

Sounds good to me. I would suggest to capture tsdump and try staging into the local roachprod to make sure that we are not seeing value doubling issue.


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
updated this, the initial version would have their metrics skipped rather than causing errors
the second commit fully supports parsing labels into DD tags

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-:-Og2mWBY0yd9Vzj7DljI:bxnm4tx)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @angles-n-daemons, and @arjunmahishi)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyJJfF79tlAXwQfs0c:-Og2mgeX3Li-1064UbYP:b-7tfmjp):*
<details><summary><i>Previously, aa-joshi (Akshay Joshi) wrote…</i></summary><blockquote>

Let's format the git message and I believe the details around querying metrics is still missing.
</blockquote></details>

added
___
*[`pkg/ts/server.go` line 587 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyKcRyAblFqat9cL1-:-Og2mid505odLqqoI2Gi:b-e964rg) ([raw file](https://github.com/cockroachdb/cockroach/blob/676c125a74776fad49a194e3c3a63cd3f34465d4/pkg/ts/server.go#L587)):*
<details><summary><i>Previously, aa-joshi (Akshay Joshi) wrote…</i></summary><blockquote>

Sounds good to me. I would suggest to capture tsdump and try staging into the local roachprod to make sure that we are not seeing value doubling issue.
</blockquote></details>

tested this, the values are ultimately split by tags, so if we upload under the same name it seems inevitable to have doubling.
I think the only workaround would be to upload to a different metric name, what do you think?


<!-- Sent from Reviewable.io -->

aa-joshi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-:-Og7fl_y-j14DNaM8UHU:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @Abhinav1299, @angles-n-daemons, @arjunmahishi, @dhartunian, and @jasonlmfong)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158716#-OfyJJfF79tlAXwQfs0c:-Og7fi0jAYqoIV0xvACb:b-yzeu17):*
<details><summary><i>Previously, jasonlmfong (Jason Fong) wrote…</i></summary><blockquote>

added
</blockquote></details>

Can you wrap the git message?


<!-- Sent from Reviewable.io -->

jasonlmfong committed:
pkg/cli: allow tsdump to also dump the labeled child metrics

This adds the ability to dump the child metrics which was introduced in #158148.
These child metrics are discovered in the tsdump through their unique resolution
and a limit list of qualified metric names.

Epic: CRDB-55079
Release: None
